CONSOLE-2394: i18n administration pages

rhamilto opened the pull request:
Note:  this PR does not include the `Administration > Namespaces` page as the code for it is shared with `Home > Projects`, and @jcaianirh has agreed to cover it as part of his story.

[CONSOLE-2394](https://issues.redhat.com/browse/CONSOLE-2394)

rhamilto commented:
/retest

spadgett commented on frontend/public/components/cluster-settings/global-config.tsx:
```diff
@@ -105,81 +113,77 @@ class GlobalConfigPage_ extends React.Component<GlobalConfigPageProps, GlobalCon
+          <div className="co-m-pane__filter-row">
+            <TextFilter
+              value={textFilter}
+              label={t('cluster-settings~by name or description')}
```
We probably need to change how this works in the component because we can't concat `Filter {label}...` as we were doing before. It needs to be one string for translation. That should be done in the common components story, though.

cc @rebeccaalpert

spadgett commented on frontend/public/components/modals/cluster-channel-modal.tsx:
```diff
@@ -13,85 +13,59 @@ import {
+            className="cluster-channel-modal__dropdown"
+            id="channel_dropdown"
+            items={availableChannels}
+            onChange={(newChannel: string) => setChannel(newChannel)}
```
```suggestion
            onChange={setChannel}
```

rhamilto replied:
Suggested change results in the following error:
`Warning: State updates from the useState() and useReducer() Hooks don't support the second callback argument. To execute a side effect after rendering, declare it in the component body with useEffect().`

I ran in to this during development and you said I needed to use the arrow function in order to remove the second parameter.

spadgett commented on frontend/public/components/modals/cluster-update-modal.tsx:
```diff
@@ -18,107 +19,77 @@ import {
-        errorMessage: `Version ${selectedVersion} not found among the available updates. Select another version.`,
-      });
+      setError(
+        `Version ${desiredVersion} not found among the available updates. Select another version.`,
```
Needs to be externalized?

rhamilto replied:
I didn't externalize this because the other error messages are coming from the API and will be in english.  My assumption is its better to keep these all the same?

spadgett commented on frontend/public/components/modals/cluster-update-modal.tsx:
```diff
@@ -18,107 +19,77 @@ import {
+            className="cluster-update-modal__dropdown"
+            id="version_dropdown"
+            items={dropdownItems}
+            onChange={(newDesiredVersion: string) => setDesiredVersion(newDesiredVersion)}
```
```suggestion
            onChange={setDesiredVersion}
```

rhamilto replied:
Same as https://github.com/openshift/console/pull/6883#discussion_r503498964

spadgett commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -371,16 +382,25 @@ const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
     <div className="co-m-pane__body co-m-pane__form">
       <Helmet>
-        <title>{`${titleVerb} Receiver`}</title>
+        <title>{`${titleVerb} ${t('alert-manager-receiver-form~Receiver')}`}</title>
```
We should avoid concatenating different translated strings where possible. It's better to make this one message.

spadgett commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -371,16 +382,25 @@ const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
-          {titleVerb} {receiverTypes[formValues.receiverType]} {isDefaultReceiver && 'Default'}{' '}
-          Receiver
+          {titleVerb} {receiverTypeLabel}
+          {isDefaultReceiver && t('alert-manager-receiver-form~Default')}{' '}
```
Same comment here. You probably want a "Default Receiver" message with the full phrase.

spadgett commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -413,15 +435,17 @@ const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
+          </label>
           <Dropdown
-            title={'Select Receiver Type...'}
+            title={t('alert-manager-receiver-form~Select Receiver Type...')}
```
Select receiver type... ?

spadgett commented on frontend/public/components/monitoring/receiver-forms/routing-labels-editor.tsx:
```diff
@@ -84,15 +89,23 @@ export const RoutingLabelEditor = ({ formValues, dispatchFormChange, isDefaultRe
-        Firing alerts with labels that match all of these selectors will be sent to this receiver.
-        Label values can be matched exactly or with a &nbsp;
-        <ExternalLink href="https://github.com/google/re2/wiki/Syntax" text="regular expression" />.
+        {t(
+          'routing-labels-editor~Firing alerts with labels that match all of these selectors will be sent to this receiver. Label values can be matched exactly or with a ',
+        )}
+        <ExternalLink
+          href="https://github.com/google/re2/wiki/Syntax"
+          text={t('routing-labels-editor~regular expression')}
+        />
+        {t('routing-labels-editor~.')}
```
Hm, I'm not sure the best way to deal with this. It really should be one message for translation. We might need to use the `Trans` component here.

spadgett commented on frontend/public/components/monitoring/receiver-forms/slack-receiver-form.tsx:
```diff
@@ -169,19 +172,19 @@ export const Form: React.FC<FormProps> = ({ globals, formValues, dispatchFormCha
                     }
                   />
                   <div className="help-block" id="slack-icon-emoji-help">
-                    An{' '}
+                    {t('slack-receiver-form~An')}{' '}
                     <ExternalLink
                       href="https://www.webfx.com/tools/emoji-cheat-sheet/"
-                      text="emoji code"
+                      text={t('slack-receiver-form~emoji code')}
                     />{' '}
-                    to use in place of the default icon.
+                    {t('slack-receiver-form~to use in place of the default icon.')}
```
Same comment here. We should see if we can make this one message.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -338,30 +283,33 @@ const Details = ({ obj: rq }) => {
             <p>
-              A cluster administrator can establish limits on both the amount you can Request and
-              your Limits with a Resource Quota.
+              {`${t(
+                'resource-quota~A cluster administrator can establish limits on both the amount you can Request and your Limits with a',
+              )} ${ResourceQuotaModel.label}${t('resource-quota~.')}`}
```
Make `label` a parameter so this can be one message

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -410,12 +414,18 @@ export const ResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
-            { id: 'namespace', title: 'Namespace Resource Quotas' },
+            {
+              id: 'cluster',
+              title: `${t('resource-quota~Cluster-wide')} ${ResourceQuotaModel.labelPlural}`,
```
This should be one message with a parameter.

spadgett reviewed (commented):
Hey @rhamilto. Really nice work on this. My main comment is that we need to be careful concatenating strings since that might not work for languages where words in a sentence might be structured in a different order. This should usually be a single message that can be translated where we plug in parameters when needed.

rhamilto commented on frontend/public/components/cluster-settings/global-config.tsx:
```diff
@@ -49,25 +35,49 @@ const ItemRow = ({ item }) => {
+  const { t } = useTranslation();
+
+  const oauthMenuItems = _.map(addIDPItems, (label: string, id: string) => ({
+    // t('oauth~Basic Authentication')
+    // t('oauth~GitHub')
+    // t('oauth~GitLab')
+    // t('oauth~Google')
+    // t('oauth~HTPasswd')
+    // t('oauth~Keystone')
+    // t('oauth~LDAP')
+    // t('oauth~OpenID Connect')
+    // t('oauth~Request Header')
```
https://github.com/openshift/console/blob/master/frontend/public/components/cluster-settings/oauth.tsx#L54-L64 is just an object.  Not sure how I'd do that.

spadgett replied:
> https://github.com/openshift/console/blob/master/frontend/public/components/cluster-settings/oauth.tsx#L54-L64 is just an object. Not sure how I'd do that.

I believe it can be anywhere in any file, and the parser will find it.

rhamilto commented on frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-details.tsx:
```diff
@@ -8,28 +9,33 @@ import {
+  const { t } = useTranslation();
+  return (
+    <div className="co-m-pane__body">
+      <SectionHeading text={`${OperatorHubModel.label} ${t('operator-hub~details')}`} />
```
@sg00dwin, FYI.  You'll want to change yours, too.

rebeccaalpert commented on frontend/public/components/limit-range.tsx:
```diff
@@ -54,49 +55,51 @@ export const LimitRangeTableRow: RowFunction<K8sResourceKind> = ({ obj, index, k
+  return (
+    <Table
+      {...props}
+      aria-label={LimitRangeModel.labelPlural}
```
Aria-labels should be translated as well; I spoke to the accessibility folks in UXD about this and they said it was important. This also applies to aria-placeholder, aria-roledescription, or aria-valuetext.

rhamilto replied:
This label is the resource name, which should not be translated.

rebeccaalpert replied:
Yup, I meant in general.

spadgett commented on frontend/public/components/cluster-settings/basicauth-idp-form.tsx:
```diff
@@ -238,3 +252,7 @@ export type AddBasicAuthPageState = {
 };
+
+type AddBasicAuthPageProps = {
+  t?: any;
```
I believe this should be `TranslationFunction` imported from `i18next`.

```suggestion
  t?: TranslationFunction;
```

But hopefully it's not needed if you can use `WithTranslation` above.

rhamilto replied:
<img width="483" alt="Screen Shot 2020-10-13 at 5 15 39 PM" src="https://user-images.githubusercontent.com/895728/95917125-bd1d1080-0d77-11eb-8235-d874d433e646.png">

spadgett replied:
It looks like @dtaylor113 was able to use `TFunction`:

https://github.com/openshift/console/pull/6928/files#diff-2c5db562a1cd010080ca3fc0f6f2277ceb8987406ab00fdda2de152470c404b9R48

spadgett commented on frontend/public/components/cluster-settings/basicauth-idp-form.tsx:
```diff
@@ -17,7 +18,10 @@ export const DroppableFileInput = (props: any) => (
 
-export class AddBasicAuthPage extends PromiseComponent<{}, AddBasicAuthPageState> {
+class AddBasicAuthPageWithTranslation extends PromiseComponent<
+  AddBasicAuthPageProps,
```
Does `WithTranslation` work here then you don't need to define your own types?

```tsx
import { withTranslation, WithTranslation } from 'react-i18next';
```
```suggestion
  WithTranslation,
```

rhamilto replied:
<img width="759" alt="Screen Shot 2020-10-13 at 5 14 08 PM" src="https://user-images.githubusercontent.com/895728/95916995-8c3cdb80-0d77-11eb-9f5a-4bc7870b3798.png">

spadgett commented on frontend/public/components/cluster-settings/cluster-settings.tsx:
```diff
@@ -612,16 +654,19 @@ export const UpdateInProgress: React.FC<UpdateInProgressProps> = ({
         <UpdatesBar>
           <Progress
-            title={`${updatedOperatorsCount} of ${totalOperatorsCount}`}
+            title={`${updatedOperatorsCount} ${t('cluster-settings~of')} ${totalOperatorsCount}`}
```
`updatedOperatorsCount` and `totalOperatorsCount` should be parameters to this message

spadgett commented on frontend/public/components/cluster-settings/cluster-settings.tsx:
```diff
@@ -631,11 +676,13 @@ export const UpdateInProgress: React.FC<UpdateInProgressProps> = ({
           <UpdatesBar>
             <Progress
-              title={`${updatedMasterNodes} of ${totalMasterNodes}`}
+              title={`${updatedMasterNodes} ${t('cluster-settings~of')} ${totalMasterNodes}`}
```
Same here, the values should be parameters to a message

spadgett commented on frontend/public/components/cluster-settings/cluster-settings.tsx:
```diff
@@ -872,35 +926,35 @@ export const ClusterVersionDetailsTable: React.FC<ClusterVersionDetailsTableProp
-  const title = 'Cluster Settings';
+export const ClusterSettingsPage: React.FC<ClusterSettingsPageProps> = ({ match }) => {
+  const { t } = useTranslation();
+  const title = t('cluster-settings~Cluster Settings');
```
Cluster settings? (Here and in the nav?)

rhamilto replied:
Page titles are supposed to be title case per http://openshift.github.io/openshift-origin-design/conventions/documentation/capitalization.html#title-case

spadgett commented on frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx:
```diff
@@ -413,15 +437,17 @@ const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
+          </label>
           <Dropdown
-            title={'Select Receiver Type...'}
+            title="Select Receiver Type..."
```
```suggestion
            title="Select receiver type..."
```

rhamilto replied:
Good catch.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/catalog-source.tsx:
```diff
@@ -372,14 +338,55 @@ const CatalogSourceTableRow: RowFunction<CatalogSourceTableRowObj> = ({
+        props: { className: tableColumnClasses[3] },
+      },
+      {
+        title: t('catalog-source~# of Operators'),
```
Oh, I see. Operators is considered a proper name, so it's capitalized.

spadgett commented:
/assign @yapei @ahardin-rh @reestr 
for QE/docs/PX review

rhamilto commented:
> /assign @yapei @ahardin-rh @reestr
> for QE/docs/PX review

hold on this until i give the all clear.  making a bunch of changes still.  :(

rhamilto commented on frontend/packages/operator-lifecycle-manager/src/components/create-catalog-source.tsx:
```diff
@@ -152,10 +181,10 @@ export const CreateCatalogSource: React.FC<CreateCatalogSourceProps> = withHandl
             <ActionGroup className="pf-c-form__group--no-top-margin">
               <Button type="submit" variant="primary" id="save-changes">
-                Create
+                {t('create-catalog-source~Create')}
```
Note there is a bug preventing the use of the `public` namespace in packages, so I omitted `Create` and `Cancel` here.  @rebeccaalpert is investigating the bug.

rhamilto commented:
/retest

rhamilto commented:
/hold cancel

I've addressed all feedback; however, I am seeing a browser console error where <Trans> is not correctly processing <strong>.  https://github.com/openshift/console/pull/6883/files#diff-fd9ef5fd9720bb43ffeb8388f6dbd76f5ec41aed6251e2ccf5ea5d34b352c743R231 results in
`Missing i18n key "Optionally list organizations. If specified, only GitHub users that are members of at least one of the listed organizations will be allowed to log in. Cannot be used in combination with <strong>teams</strong>." in namespace "github-idp-form" and language "en." `
but should be looking for https://github.com/openshift/console/pull/6883/files#diff-d2e7e429f9c33c426bb8916cc4bb03f9cf02d6c91a3aedbf04af41c3679f12a1R10

yanpzhan commented:
@rhamilto  "More Options" in openid-idp-form is not marked for translation, should it be marked?

rhamilto commented:
> @rhamilto "More Options" in openid-idp-form is not marked for translation, should it be marked?

Good catch.  That is in a shared component (`<ListInput>`), which we've said we're going to tackle separately.  So my assumption is that we'll get it in a separate PR.

rhamilto commented:
> /hold cancel
> 
> I've addressed all feedback; however, I am seeing a browser console error where `<Trans>` is not correctly processing `<strong>`. https://github.com/openshift/console/pull/6883/files#diff-fd9ef5fd9720bb43ffeb8388f6dbd76f5ec41aed6251e2ccf5ea5d34b352c743R231 results in
> `Missing i18n key "Optionally list organizations. If specified, only GitHub users that are members of at least one of the listed organizations will be allowed to log in. Cannot be used in combination with <strong>teams</strong>." in namespace "github-idp-form" and language "en." `
> but should be looking for https://github.com/openshift/console/pull/6883/files#diff-d2e7e429f9c33c426bb8916cc4bb03f9cf02d6c91a3aedbf04af41c3679f12a1R10

I found a workaround for this issue by [adding an `i18nKey` prop to `<Trans>`](https://github.com/openshift/console/pull/6883/files#diff-fd9ef5fd9720bb43ffeb8388f6dbd76f5ec41aed6251e2ccf5ea5d34b352c743R228).

@spadgett, PR is ready for final review.

rhamilto commented:
/retest

rhamilto commented:
/retest

rebeccaalpert commented:
You don't need to add a key; that'll break our existing pattern for keys. It looks like you're missing the `t` prop for the Trans component. You need to do something like `<Trans t={t} ns="namespace">...</Trans>`.

rhamilto commented:
/retest

spadgett reviewed (approved):
/approve

ahardin-rh commented on frontend/public/components/cluster-settings/ldap-idp-form.tsx:
```diff
@@ -249,46 +256,56 @@ export class AddLDAPPage extends PromiseComponent<{}, AddLDAPPageState> {
             onChange={this.attributesPreferredUsernameChanged}
-            helpText="The list of attributes whose values should be used as the preferred username."
+            helpText={t(
+              'ldap-idp-form~The list of attributes whose values should be used as the preferred username.',
```
In docs, we use "user name" vs "username" per IBM style guidelines. Probably out of scope for this PR, but just a heads-up in case we want to sync up our usage on that.

ahardin-rh commented on frontend/public/components/monitoring/receiver-forms/email-receiver-form.tsx:
```diff
@@ -136,7 +141,7 @@ export const Form: React.FC<FormProps> = ({ globals, formValues, dispatchFormCha
                 />
                 <div className="help-block" id="email-hello-help">
-                  The hostname to identify to the SMTP server.
+                  {t('email-receiver-form~The hostname to identify to the SMTP server.')}
```
In docs, we use host name per the IBM style guide. Another likely out of scope comment, but something we might consider syncing up on later :smile:

ahardin-rh commented on frontend/public/components/monitoring/receiver-forms/pagerduty-receiver-form.tsx:
```diff
@@ -113,33 +117,34 @@ export const Form: React.FC<FormProps> = ({ globals, formValues, dispatchFormCha
-              tooltip="Checking this box will write the url to the global section of the
-                configuration file where it will become default url for future PagerDuty receivers."
+              tooltip={t(
+                'pagerduty-receiver-form~Checking this box will write the url to the global section of the configuration file where it will become default url for future PagerDuty receivers.',
```
s/url/URL

ahardin-rh commented:
Ack from the docs side :+1:

yanpzhan commented:
/label qe-approved

spadgett commented:
> Ack from the docs side 👍

/label docs-approved

rhamilto commented:
/hold for px-approval

spadgett reviewed (approved):
/lgtm

spadgett commented:
/unassign @reestr 
/assign @sferich888

sferich888 commented:
/label px-approved

spadgett commented:
/hold cancel
